datafusion/physical-plan/src/tree_node.rs

```diff
@@ -42,6 +42,8 @@ impl DynTreeNode for dyn ExecutionPlan {
 /// A node object beneficial for writing optimizer rules, encapsulating an [`ExecutionPlan`] node with a payload.
 /// Since there are two ways to access child plans—directly from the plan and through child nodes—it's recommended
 /// to perform mutable operations via [`Self::update_plan_from_children`].
+/// After update `children`, please do the sync updating for `plan`'s children.
```

Comment:
I wonder if we could find some way to avoid having to remember to do this 🤔 

If we have to remember to call a function it seems likely either our future selves or others will forget to do so

Reply:
Maybe one way would be to make these fields not `pub` and then control access

so like add a method so the only way to update children woudl be

```rust
impl PlanContext {
  fn update_children(&mut self, new_children: Vec<Self>) -> {
    self.children = new_children;
    self.update_child_data()
}
```
🤔 

To be clear I am talking about some future PR

Reply:
Yes, I've tried this way to make the operation "atomic", but I found it hard to change, most places acquire `mut self` for `PlanContext`.

I have another way to avoid the mind burden. Currrently, we have two parallel tree structures:
- The ExecutionPlan tree (through plan field and its children)
- The PlanContext tree (through children field)

I'm trying to make them single.